primitives/allocator/src/freeing_bump.rs

```diff
@@ -93,6 +81,219 @@ macro_rules! trace {
 	}
 }
 
+/// The exponent for the power of two sized block adjusted to the minimum size.
+///
+/// This way, if MIN_POSSIBLE_ALLOCATION == 8, we would get:
+///
+/// power_of_two_size | order
+/// 8                 | 0
+/// 16                | 1
+/// 32                | 2
+/// 64                | 3
+///
+/// and so on.
+#[derive(Copy, Clone, PartialEq, Eq, Debug)]
+struct Order(u32);
+
+impl Order {
+	/// Create `Order` object from a raw order.
+	///
+	/// Returns `Err` if it is greater than the maximum supported order.
+	fn from_raw(order: u32) -> Result<Self, Error> {
+		if order < N as u32 {
+			Ok(Self(order))
+		} else {
+			Err(error("invalid order"))
+		}
+	}
+
+	/// Compute the order by the given size
+	///
+	/// The size is clamped, so that the following holds:
+	//
+	// MIN_POSSIBLE_ALLOCATION <= size <= MAX_POSSIBLE_ALLOCATION
+	fn from_size(size: u32) -> Result<Self, Error> {
+		let clamped_size = if size > MAX_POSSIBLE_ALLOCATION {
+			return Err(Error::RequestedAllocationTooLarge);
+		} else if size < MIN_POSSIBLE_ALLOCATION {
+			MIN_POSSIBLE_ALLOCATION
+		} else {
+			size
+		};
+
+		// Round the clamped size to the next power of two.
+		//
+		// It returns the unchanged value if the value is already a power of two.
+		let power_of_two_size = clamped_size.next_power_of_two();
+
+		// Compute the number of trailing zeroes to get the order. We adjust it by the number of
+		// trailing zeroes in the minimum possible allocation.
+		let order = power_of_two_size.trailing_zeros() - MIN_POSSIBLE_ALLOCATION.trailing_zeros();
+
+		Ok(Self(order))
+	}
+
+	/// Returns the corresponding size for this order.
+	///
+	/// Note that it is always a power of two.
+	fn size(&self) -> u32 {
+		1 << MIN_POSSIBLE_ALLOCATION.trailing_zeros() << self.0
+	}
+
+	/// Extract the order as `u32`.
+	fn into_raw(self) -> u32 {
+		self.0
+	}
+}
+
+/// A marker for denoting the end of the linked list.
+const EMPTY_MARKER: u32 = u32::max_value();
+
+/// A link between headers in the free list.
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+enum Link {
+	/// Null, denotes that there is no next element.
+	Null,
+	/// Link to the next element represented as a pointer to the a header.
+	Ptr(u32),
+}
+
+impl Link {
+	/// Creates a link from raw value.
+	fn from_raw(raw: u32) -> Self {
+		if raw != EMPTY_MARKER {
+			Self::Ptr(raw as u32)
+		} else {
+			Self::Null
+		}
+	}
+
+	/// Converts this link into a raw u32.
+	fn into_raw(self) -> u32 {
+		match self {
+			Self::Null => EMPTY_MARKER,
+			Self::Ptr(ptr) => ptr,
+		}
+	}
+}
+
+/// A header of an allocation.
+///
+/// The header is encoded in memory as follows.
+///
+/// ## Free header
+///
+/// ```ignore
+/// 64             32                  0
+//  +--------------+-------------------+
+/// |            0 | next element link |
```

Comment:
doesn't this end up using twice as much memory for the header?

Reply:
Compared to the former version? No, before the refactoring the header (né `PREFIX`) was also 8 bytes but was only using only the lower 32 bits. In this PR we just start using the upper ones.

That said, I think there are definitely an area of improvement. For example we can leverage the fact that we store either an order, which doesn't require a lot of bits, and a pointer, which is aligned to 8 therefore we can make use of 3 lower bits, thus decreasing the minimal size required for the header to 32 bits without a lot of hassle (we would need to preserve alignment guarantees though).

I didn't do anything alike to keep this PR simple and to make it easier to review.